arch/risc-v/src/common/espressif/esp_qencoder.c

```diff
@@ -0,0 +1,593 @@
+  pcnt_ll_set_low_limit_value(priv->dev, priv->config->pcntid, INT16_MIN);
+
+  /* Setup POS/NEG/LCTRL/HCTRL/FILTER modes */
+
+  pcnt_ll_set_glitch_filter_thres(priv->dev, priv->config->pcntid, priv->config->filter_thres);
+  pcnt_ll_enable_glitch_filter(priv->dev, priv->config->pcntid, true);
+  pcnt_ll_set_edge_action(priv->dev, priv->config->pcntid, 0, PCNT_CHANNEL_EDGE_ACTION_DECREASE, PCNT_CHANNEL_EDGE_ACTION_INCREASE);
+  pcnt_ll_set_level_action(priv->dev, priv->config->pcntid, 0, PCNT_CHANNEL_LEVEL_ACTION_KEEP, PCNT_CHANNEL_LEVEL_ACTION_INVERSE);
+
+  pcnt_ll_set_edge_action(priv->dev, priv->config->pcntid, 1, PCNT_CHANNEL_EDGE_ACTION_INCREASE, PCNT_CHANNEL_EDGE_ACTION_DECREASE);
+  pcnt_ll_set_level_action(priv->dev, priv->config->pcntid, 1, PCNT_CHANNEL_LEVEL_ACTION_KEEP, PCNT_CHANNEL_LEVEL_ACTION_INVERSE);
+
+  /* Configure GPIO pins as Input with Pull-Up enabled */
+
+  esp_configgpio(priv->config->ch0_gpio, INPUT_FUNCTION_3 | PULLUP);
+  esp_configgpio(priv->config->ch1_gpio, INPUT_FUNCTION_3 | PULLUP);
+
+  /* Connect Channel A (ch0_gpio) and Channel B (ch1_gpio) crossed for X4 */
+
+  esp_gpio_matrix_in(priv->config->ch0_gpio,
+                     priv->config->ch0_pulse_sig, 0);
+  esp_gpio_matrix_in(priv->config->ch1_gpio,
+                     priv->config->ch0_ctrl_sig, 0);
+
+  esp_gpio_matrix_in(priv->config->ch1_gpio,
+                     priv->config->ch1_pulse_sig, 0);
+  esp_gpio_matrix_in(priv->config->ch0_gpio,
+                     priv->config->ch1_ctrl_sig, 0);
+
+  /* Clear the Reset bit to enable the Pulse Counter */
+
+  pcnt_ll_clear_count(priv->dev, priv->config->pcntid);
+  pcnt_ll_start_count(priv->dev, priv->config->pcntid);
+
+  spin_unlock_irqrestore(&priv->lock, flags);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp_shutdown
+ *
+ * Description:
+ *   This method is called when the driver is closed.  The lower half driver
+ *   should stop data collection, free any resources, disable timer hardware,
+ *   and put the system into the lowest possible power usage state
+ *
+ ****************************************************************************/
+
+static int esp_shutdown(struct qe_lowerhalf_s *lower)
+{
+  struct esp_lowerhalf_s *priv = (struct esp_lowerhalf_s *)lower;
+
+  /* Disable PCNT clock */
+
+  periph_module_disable(PERIPH_PCNT_MODULE);
+
+  /* Make sure initial position is 0 */
+
+  priv->position = 0;
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp_position
+ *
+ * Description:
+ *   Return the current position measurement.
+ *
+ ****************************************************************************/
+
+static int esp_position(struct qe_lowerhalf_s *lower, int32_t *pos)
+{
+  struct esp_lowerhalf_s *priv = (struct esp_lowerhalf_s *)lower;
+  irqstate_t flags;
+  int32_t position;
+  int16_t count;
+
+  DEBUGASSERT(lower && priv->inuse);
```

Comment:
You can move DEBUGASSERT before setting `*priv`

Reply:
done